### PR TITLE
a8n: Align preview and campaign logic

### PR DESCRIPTION
**Pull request by eseliger**
This increases the overlap of logic for preview and campaigns
ChangesetPlans are loaded in pages, like Changesets were already

note: the large diff is only because the test coverage was also increased

### CODE REVIEW

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 1]
```diff
@@ -589,7 +589,7 @@ type Campaign implements Node {
     plan: CampaignPlan
 
     # The current status of creating the campaigns changesets on the code host.
-    changesetCreationStatus: BackgroundProcessStatus
+    changesetCreationStatus: BackgroundProcessStatus!
```
Comment (eseliger) [inline cmd/frontend/graphqlbackend/schema.graphql]: @sourcegraph/automation it seems in the code this is never null, and that removed an edge-case to handle so I made it non-optional. Can you confirm this is correct?

Comment (mrnugget) [reply cmd/frontend/graphqlbackend/schema.graphql]: Right now it's always non-null, yes.

Comment (felixfbecker) [reply cmd/frontend/graphqlbackend/schema.graphql]: If the campaign has no plan, what is the background process status? 🤔

Comment (eseliger) [reply cmd/frontend/graphqlbackend/schema.graphql]: "nothing to do, done"

File: web/src/enterprise/campaigns/detail/CampaignDetails.tsx [hunk 2]
```diff
@@ -195,18 +196,42 @@ export const CampaignDetails: React.FunctionComponent<Props> = ({
...
     useEffect(() => {
         if (planID) {
             nextPreviewCampaignPlan(planID)
         }
     }, [nextPreviewCampaignPlan, planID])
 
+    const status = campaign
+        ? campaign.__typename === 'CampaignPlan'
+            ? campaign.status
+            : campaign.changesetCreationStatus
```
Comment (felixfbecker) [inline web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: @mrnugget @ryanslade would it be possible to make these field names consistent (as if both types were implementing an interface)?

File: web/src/enterprise/campaigns/detail/FileDiffTabNode.tsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    location,
+    isLightTheme,
+}) =>
+    node.diff && (
+        <div>
+            <h3>
+                <SourcePullIcon className="icon-inline mr-2" />{' '}
+                <LinkOrSpan to={node.repository.url}>{node.repository.name}</LinkOrSpan>
+            </h3>
+            {((): (GQL.IPreviewFileDiff | GQL.IFileDiff)[] => node.diff.fileDiffs.nodes)().map(fileDiffNode => (
```
Comment (felixfbecker) [inline web/src/enterprise/campaigns/detail/FileDiffTabNode.tsx]: smart

File: web/src/enterprise/campaigns/detail/presentation.ts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+/** The maximum amount of changeset nodes rendered initially */
+export const DEFAULT_LIST_COUNT = 15
```
Comment (felixfbecker) [inline web/src/enterprise/campaigns/detail/presentation.ts]: Would recommend renaming to DEFAULT_CHANGESET_LIST_COUNT so it's globally unique

File: web/src/enterprise/campaigns/detail/CampaignDetails.tsx (hunk 2 above)
Comment (mrnugget) [reply web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: Absolutely: https://github.com/sourcegraph/sourcegraph/pull/7654 :)

Feel free to jump on that branch and edit the TypeScript etc.

**Pull request merged**